From a3a7f90bc5dad7c9a41b4e907eb5f0f38105e93f Mon Sep 17 00:00:00 2001
From: Robin Gong <yibin.gong@nxp.com>
Date: Thu, 6 Sep 2018 23:41:43 +0800
Subject: [PATCH 4554/5242] MLK-19418-2: dmaengine: imx-sdma: make sure
 firmware loaded

commit  5481ee381e94e017135334d44c6936590d3930b2 from
https://source.codeaurora.org/external/imx/linux-imx.git

request_firmware_nowait() is async so that can't make sure sdma firmware
loaded when sdma_resume return back. Add wait code to make sure firmware
ready.

Signed-off-by: Robin Gong <yibin.gong@nxp.com>
Reviewed-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/imx-sdma.c |    9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index 25c43d3..d59eb26 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -2544,6 +2544,7 @@ static int sdma_resume(struct device *dev)
 {
 	struct platform_device *pdev = to_platform_device(dev);
 	struct sdma_engine *sdma = platform_get_drvdata(pdev);
+	unsigned long timeout = jiffies + msecs_to_jiffies(2);
 	int i, ret;
 
 	/* Do nothing if not i.MX6SX or i.MX7D*/
@@ -2589,6 +2590,14 @@ static int sdma_resume(struct device *dev)
 		dev_warn(&pdev->dev, "failed to get firware\n");
 		goto out;
 	}
+	/* wait firmware loaded */
+	do {
+		if (time_after(jiffies, timeout)) {
+			dev_warn(&pdev->dev, "failed to load firmware\n");
+			break;
+		}
+		usleep_range(50, 500);
+	} while (!sdma->fw_loaded);
 
 	ret = sdma_save_restore_context(sdma, false);
 	if (ret) {
-- 
1.7.9.5

